### PR TITLE
Stability fixes

### PR DESCRIPTION
**Pull request by lower29**
This pull request fixes a bunch of different stuff, but bear with me - don't hit the "reject" button just yet!

Here are the included changes:

1. Use a single Docker volume for persistence of data. The base image has five volumes - this is too many to keep track of! We now put all installed data and state under our own `/var/lib/nextcloud` volume.
1. Fix capturing of log output. We were capturing the output from the run and setup, but NextCloud, PHP and Nginx logs were all still being written to file. They are now streamed to `stdout`, where `docker logs` can pick them up from. For complicated reasons this is done using FIFOs rather than linking or writing to `/dev/stdout` - it seems PHP doesn't behave well when writing to devices rather than files so we have to fool it.
1. Fix caching. Once I could see the logs properly, I could see NextCloud was complaining about not being able to use APCu, which it tries to use for its Memcache implementation. Sure enough APCu wasn't enabled in PHP - now it is. I also added a depedency on Redis too, because this further improves caching performance - or is supposed to anyway.
1. Admin user has "example" files in its data folder. The admin user gets created by the installer, and previously this meant that the installer was copying the "skeleton" files into admin's home space, even though it wasn't for other users. This is now fixed - the installer doesn't copy any files.
1. The user/group NextCloud runs as doesn't exist. We specify a `UID` and `GID` as environment variables but no user was being created. This is now done as part of the initial boot of the container, in the `run.sh` script (which we've overridden from the base image).

The most important of these is the consolidation of disparate data file locations into `/var/lib/nextcloud`. This means we can now persist **all** NextCloud files required for correct operation across restarts, and can stop using workarounds to restore data and configuration that was inadvertently lost.

Hopefully with all of these changes in place we should be on the way towards a more stable deployment.

@sevein Can you please review? I realise you don't have a great deal of experience with NextCloud (or any?) but a lot of this is more to do with Docker than NextCloud, so your feedback is definitely welcome. @jeremysparks86 If you get the chance to review too then that would be great. Thanks.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -24,6 +19,13 @@ services:
         expose:
             - "3306"
 
+    redis:
```
Comment (lower29) [inline docker-compose.yml]: This service is the same as is used in [rdss-archivematica](https://github.com/JiscRDSS/rdss-archivematica/blob/master/compose/docker-compose.qa.yml), which is why the persistence has been disabled. I wanted to be able to reuse the Redis service defined there, even though this Compose config is only for dev and example purposes.

**Comment by lower29** [comment]
Oops, I'll fix the linting errors.

The rationale behind using APCu and Redis is that they're being used for different things: APCu is used to cache rendered web pages and other assets, Redis is used to cache file locks and database queries. The [OwnCloud configuration documentation](https://doc.owncloud.org/server/10.0/admin_manual/configuration/server/caching_configuration.html#small-organization-single-server-setup) has some discussion on the merits of the different caching mechanisms, based on the discussion and benchmarking done [here](https://github.com/owncloud/documentation/pull/1992#issuecomment-203554416) - see [this PR](https://github.com/owncloudarchive/pi-image/issues/24) for the raw numbers.

The config documentation considers three typical deployment classes:

1. Small/Private Home Server - only use APCu
1. Small Organization, Single-Server Setup - use APCu for cache and Redis for locking
1. Large Organization, Clustered Setup - use Redis for everything except local memory cache

In terms of our configuration, I'm characterising the deployment environment as container-based, which means each container will have minimal resources but other containers can be asked to do the work. We can also anticipate that each deployment will have a lot of work to do (e.g. the York dataset is just the tip of the iceberg). This means that the "Small/Private Home Server" is not appropriate: APCu stores everything in memory, which our container should not be assumed to have a lot of. "Small Organization, Single-Server Setup" also doesn't fit, because we are expecting this to scale, which doesn't scream "Small Organization". This leaves us with the "Large Organization", i.e. Redis for everything except local cache, which APCu mops up.

One "flaw" in the current configuration template is that it assumes a single redis host. In future we may need to change this, i.e. when we are using `docker-compose scale redis=2` or deploying with Amazon RDS. However, for now it's fine - we can leave that for another PR.

**Comment by lower29** [comment]
Linting errors fixed in rebase. The CI is now passing, so I'm going to go ahead and merge.

**Comment by sevein** [comment]
Thanks for the explanation.

**Pull request merged**